main.go

```diff
@@ -0,0 +1,367 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+// parseQueryParams returns the start and end values, or an error.
+func parseQueryParams(values url.Values) (int64, int64, error) {
+	start := values.Get("start")
+	end := values.Get("end")
+	if start == "" {
+		return 0, 0, errors.New("missing start parameter")
+	}
+	if end == "" {
+		return 0, 0, errors.New("missing end parameter")
+	}
+	startInt, err := strconv.ParseInt(start, 10, 64)
+	if err != nil || startInt < 0 {
+		return 0, 0, errors.New("invalid start parameter")
+	}
+	endInt, err := strconv.ParseInt(end, 10, 64)
+	if err != nil || endInt < 0 {
+		return 0, 0, errors.New("invalid end parameter")
+	}
+	if endInt <= startInt {
+		return 0, 0, errors.New("end must be greater than start")
+	}
+	return startInt, endInt, nil
+}
+
+// tile represents important numbers about a tile: where it starts, where it ends, its size,
+// and what CT backend URL it exists on (or is anticipated to exist on), and what s3 prefix
+// it should be stored/retrieved under.
+type tile struct {
+	start    int64
+	end      int64
+	size     int64
+	logURL   string
+	s3prefix string
```

Comment:
It feels a little weird to me that a tile carries these two pieces of data (which are identical for all tiles within a single log), but doesn't carry the s3 bucket or the API path under that log hostname, which are also identical for all tiles within a single log.

It results in odd function signatures below, where both getFromS3 and writeToS3 take a bucket name but then rely on the tile for the prefix, despite both of those ultimately coming from the same CLI flags.

I don't think I'm advocating for removing these struct memebers -- replacing them with function args everywhere would be annoyingly verbose. So maybe the s3bucket should also be a struct member here? Or maybe things are fine as they are. Just worth thinking about.

Reply:
yeah I wound up moving the helpers onto the handler struct. good idea. 👍🏻

that allowed me to remove the `s3prefix` from `tile`, which was nice. I felt a little awkward having `tile` represent both data about the CT log and about the S3 storage, like it was trying to be in two worlds at once. Now it just contains data about the CT log, which is nice.